src/vos/vos_dtx.c

```diff
@@ -717,6 +730,469 @@ vos_dtx_abort_one(struct vos_container *cont, struct daos_tx_id *dti,
+
+			obj = umem_id2ptr(umm, vds->vds_record);
+
+			/* The to be shared obj has been committed. */
+			if (UMMID_IS_NULL(obj->vo_dtx))
+				goto drop;
+
+			umem_tx_add_ptr(umm, obj, sizeof(*obj));
+
+			/* The to be shared obj has been aborted, reuse it. */
+			if (umem_id_equal(umm, obj->vo_dtx,
+					  DTX_UMMID_ABORTED)) {
+				D_ASSERTF(obj->vo_dtx_shares == 0,
+					  "Invalid shared obj DTX shares %d\n",
+					  obj->vo_dtx_shares);
+				obj->vo_dtx_shares = 1;
+				goto add;
+			}
+
+			obj->vo_dtx_shares++;
+			shared = true;
+
+			/* The original obj DTX has been aborted, but some
+			 * others still share the obj. Then set the vo_dtx
+			 * to current DTX.
+			 */
+			if (umem_id_equal(umm, obj->vo_dtx,
+					  DTX_UMMID_UNKNOWN)) {
+				obj->vo_dtx = dth->dth_ent;
+
+				D_DEBUG(DB_TRACE, "The DTX "DF_DTI" shares obj "
+					"with unknown DTXs, shares count %u.\n",
+					DP_DTI(&dth->dth_xid),
+					obj->vo_dtx_shares);
+				goto add;
+			}
+
+			sh_dtx = umem_id2ptr(umm, obj->vo_dtx);
+			D_ASSERT(dtx != sh_dtx);
+			D_ASSERTF(sh_dtx->te_state == DTX_ST_PREPARED ||
+				  sh_dtx->te_state == DTX_ST_INIT,
+				  "Invalid shared obj DTX state: %u\n",
+				  sh_dtx->te_state);
+
+			umem_tx_add_ptr(umm, sh_dtx, sizeof(*sh_dtx));
+			sh_dtx->te_flags |= DTX_EF_SHARES;
+
+			D_DEBUG(DB_TRACE, "The DTX "DF_DTI" shares obj with "
+				DF_DTI", the shares count %u.\n",
+				DP_DTI(&dth->dth_xid), DP_DTI(&sh_dtx->te_xid),
+				obj->vo_dtx_shares);
+		} else {
+			struct vos_krec_df	*key;
+
+			key = umem_id2ptr(umm, vds->vds_record);
+
+			/* The to be shared key has been committed. */
+			if (UMMID_IS_NULL(key->kr_dtx))
+				goto drop;
+
+			umem_tx_add_ptr(umm, key, sizeof(*key));
+
+			/* The to be shared key has been aborted, reuse it. */
+			if (umem_id_equal(umm, key->kr_dtx,
+					  DTX_UMMID_ABORTED)) {
+				D_ASSERTF(key->kr_dtx_shares == 0,
+					  "Invalid shared key DTX shares %d\n",
+					  key->kr_dtx_shares);
+				key->kr_dtx_shares = 1;
+				goto add;
+			}
+
+			key->kr_dtx_shares++;
+			shared = true;
+
+			/* The original key DTX has been aborted, but some
+			 * others still share the key. Then set the kr_dtx
+			 * to current DTX.
+			 */
+			if (umem_id_equal(umm, key->kr_dtx,
+					  DTX_UMMID_UNKNOWN)) {
+				key->kr_dtx = dth->dth_ent;
+
+				D_DEBUG(DB_TRACE, "The DTX "DF_DTI" shares key "
+					"with unknown DTXs, shares count %u.\n",
+					DP_DTI(&dth->dth_xid),
+					key->kr_dtx_shares);
+				goto add;
+			}
+
+			sh_dtx = umem_id2ptr(umm, key->kr_dtx);
+			D_ASSERT(dtx != sh_dtx);
+			D_ASSERTF(sh_dtx->te_state == DTX_ST_PREPARED ||
+				  sh_dtx->te_state == DTX_ST_INIT,
+				  "Invalid shared key DTX state: %u\n",
+				  sh_dtx->te_state);
+
+			umem_tx_add_ptr(umm, sh_dtx, sizeof(*sh_dtx));
+			sh_dtx->te_flags |= DTX_EF_SHARES;
+
+			D_DEBUG(DB_TRACE, "The DTX "DF_DTI" shares key with "
+				DF_DTI", the shares count %u.\n",
+				DP_DTI(&dth->dth_xid), DP_DTI(&sh_dtx->te_xid),
+				key->kr_dtx_shares);
+		}
```

Comment:
Style: better to reorganize above code to avoid code duplication of obj and key processing

Reply:
I will move the debug() code outside of the condition check. But because object and key have different flags, so they need to be distinguished and handled differently.